dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/predictor/PredictorDeletionHandler.java

```diff
@@ -111,6 +113,15 @@ private DeletionVeto allowDeleteCategoryOptionCombo( CategoryOptionCombo optionC
         return vetoIfExists( "SELECT COUNT(*) FROM predictor where generatoroutputcombo=" + optionCombo.getId() );
     }
 
+    private DeletionVeto allowDeleteCategoryCombo( CategoryCombo categoryCombo )
+    {
+        return vetoIfExists( "SELECT COUNT(*) FROM predictor p where exists ("
```

Comment:
I am wondering if we are not fixing the root cause here. It should be enough to check for the category option combo of the predictor. 

The question is why PredictorDeletionHandler#allowDeleteCategoryOptionCombo is not triggered. This check should be triggered when the delete of category option combo is attempted.

Reply:
Yeah, we talked about this. It should have worked without this extra check but due to order or manner of how this is processed it does not work as we want it to. This could be considered a workaround. In general I think we could need something more centralised to deal with the blocking or unlinking on deletion. While the deletion handler concept is nicely composable there appear to be difficulties arising from the looseness of the concept that for example gives little to no control to order or combinations.

Reply:
Just a guess, but i think `PredictorDeletionHandler#allowDeleteCategoryOptionCombo` is not triggered because the deletion of `CategoryOptionCombo` occurs in this `updateNoAcl` method in class `CategoryComboDeletionHandler`

```java
private void deleteCategoryOptionCombo( CategoryOptionCombo categoryOptionCombo )
    {
        for ( CategoryCombo categoryCombo : categoryService.getAllCategoryCombos() )
        {
            categoryCombo.getOptionCombos().remove( categoryOptionCombo );
            idObjectManager.updateNoAcl( categoryCombo );
        }
    }
```

Reply:
I think this is not actually where the option combo is deleted @vietnguyen , I think this is where we remove the association from category combo before deleting the option combo.